File: frontend/utils/dataValidation.ts
/**
 * Frontend data validation utilities.
 *
 * Provides utilities to validate data structures received from backend,
 * compare expected vs actual data, and report mismatches.
 */

import { logWebSocketMessage } from './communicationLogger';

export interface DataMismatch {
  fieldPath: string;
  expectedValue: any;
  actualValue: any;
  mismatchType: 'type' | 'value' | 'missing' | 'extra' | 'structure';
  description: string;
}

export interface ValidationResult {
  isValid: boolean;
  mismatches: DataMismatch[];
  validatedAt: string;
}

export class DataValidator {
  private mismatches: DataMismatch[] = [];
  private strictMode: boolean;

  constructor(strictMode: boolean = false) {
    this.strictMode = strictMode;
  }

  /**
   * Validate data structure against expected schema
   */
  validateStructure(
    data: any,
    expectedSchema: any,
    path: string = ''
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    if (expectedSchema.type === 'object') {
      mismatches.push(...this.validateObjectStructure(data, expectedSchema, path));
    } else if (expectedSchema.type === 'array') {
      mismatches.push(...this.validateArrayStructure(data, expectedSchema, path));
    } else {
      mismatches.push(...this.validatePrimitive(data, expectedSchema, path));
    }

    // Check required fields
    if (expectedSchema.required && (data === null || data === undefined)) {
      mismatches.push({
        fieldPath: path,
        expectedValue: 'non-null value',
        actualValue: null,
        mismatchType: 'missing',
        description: 'Required field is null or missing'
      });
    }

    this.mismatches.push(...mismatches);
    return mismatches;
  }

  private validateObjectStructure(
    data: any,
    schema: any,
    path: string
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    if (typeof data !== 'object' || data === null || Array.isArray(data)) {
      return [{
        fieldPath: path,
        expectedValue: 'object',
        actualValue: this.getTypeName(data),
        mismatchType: 'type',
        description: 'Expected object but got different type'
      }];
    }

    const expectedProperties = schema.properties || {};

    // Check for missing required properties
    const requiredProps = schema.required || [];
    for (const prop of requiredProps) {
      if (!(prop in data)) {
        mismatches.push({
          fieldPath: path ? `${path}.${prop}` : prop,
          expectedValue: 'present',
          actualValue: 'missing',
          mismatchType: 'missing',
          description: `Required property '${prop}' is missing`
        });
      }
    }

    // Check existing properties
    for (const [prop, propSchema] of Object.entries(expectedProperties)) {
      const propPath = path ? `${path}.${prop}` : prop;
      if (prop in data) {
        mismatches.push(...this.validateStructure(data[prop], propSchema, propPath));
      } else if ((propSchema as any).required) {
        mismatches.push({
          fieldPath: propPath,
          expectedValue: 'present',
          actualValue: 'missing',
          mismatchType: 'missing',
          description: `Required property '${prop}' is missing`
        });
      }
    }

    // Check for extra properties if schema is strict
    if (schema.additionalProperties === false) {
      const allowedProps = new Set(Object.keys(expectedProperties));
      const actualProps = new Set(Object.keys(data));
      const extraProps = Array.from(actualProps).filter(prop => !allowedProps.has(prop));

      for (const extraProp of extraProps) {
        const propPath = path ? `${path}.${extraProp}` : extraProp;
        mismatches.push({
          fieldPath: propPath,
          expectedValue: 'not present',
          actualValue: data[extraProp],
          mismatchType: 'extra',
          description: `Extra property '${extraProp}' not allowed by schema`
        });
      }
    }

    return mismatches;
  }

  private validateArrayStructure(
    data: any,
    schema: any,
    path: string
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    if (!Array.isArray(data)) {
      return [{
        fieldPath: path,
        expectedValue: 'array',
        actualValue: this.getTypeName(data),
        mismatchType: 'type',
        description: 'Expected array but got different type'
      }];
    }

    const itemSchema = schema.items;

    // Validate array length constraints
    const minItems = schema.minItems;
    const maxItems = schema.maxItems;

    if (minItems !== undefined && data.length < minItems) {
      mismatches.push({
        fieldPath: path,
        expectedValue: `at least ${minItems} items`,
        actualValue: data.length,
        mismatchType: 'value',
        description: `Array has ${data.length} items but minimum is ${minItems}`
      });
    }

    if (maxItems !== undefined && data.length > maxItems) {
      mismatches.push({
        fieldPath: path,
        expectedValue: `at most ${maxItems} items`,
        actualValue: data.length,
        mismatchType: 'value',
        description: `Array has ${data.length} items but maximum is ${maxItems}`
      });
    }

    // Validate each item
    if (itemSchema) {
      data.forEach((item, index) => {
        const itemPath = `${path}[${index}]`;
        mismatches.push(...this.validateStructure(item, itemSchema, itemPath));
      });
    }

    return mismatches;
  }

  private validatePrimitive(
    data: any,
    schema: any,
    path: string
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    const expectedType = schema.type;
    if (expectedType) {
      const actualType = this.getJsonType(data);

      // Allow numeric types to be flexible
      if (expectedType === 'number') {
        if (actualType !== 'number' && actualType !== 'integer' && actualType !== 'float') {
          mismatches.push({
            fieldPath: path,
            expectedValue: expectedType,
            actualValue: actualType,
            mismatchType: 'type',
            description: `Expected ${expectedType} but got ${actualType}`
          });
        }
      } else if (expectedType !== actualType) {
        mismatches.push({
          fieldPath: path,
          expectedValue: expectedType,
          actualValue: actualType,
          mismatchType: 'type',
          description: `Expected ${expectedType} but got ${actualType}`
        });
      }
    }

    // Validate enum values
    const enumValues = schema.enum;
    if (enumValues && !enumValues.includes(data)) {
      mismatches.push({
        fieldPath: path,
        expectedValue: enumValues,
        actualValue: data,
        mismatchType: 'value',
        description: `Value ${data} not in allowed enum values ${enumValues}`
      });
    }

    return mismatches;
  }

  private getJsonType(value: any): string {
    if (value === null) return 'null';
    if (typeof value === 'boolean') return 'boolean';
    if (typeof value === 'number') return Number.isInteger(value) ? 'integer' : 'number';
    if (typeof value === 'string') return 'string';
    if (Array.isArray(value)) return 'array';
    if (typeof value === 'object') return 'object';
    return typeof value;
  }

  private getTypeName(value: any): string {
    if (value === null) return 'null';
    if (Array.isArray(value)) return 'array';
    return typeof value;
  }

  /**
   * Compare expected vs actual data for equality
   */
  compareData(
    expected: any,
    actual: any,
    path: string = '',
    tolerance: number = 0
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    // Handle null values
    if (expected === null && actual === null) return mismatches;
    if (expected === null || actual === null) {
      return [{
        fieldPath: path,
        expectedValue: expected,
        actualValue: actual,
        mismatchType: 'value',
        description: 'One value is null, other is not'
      }];
    }

    // Handle different types
    if (typeof expected !== typeof actual) {
      // Special case: allow number/string interchange for numeric values
      if (typeof expected === 'number' && typeof actual === 'string') {
        const parsed = parseFloat(actual);
        if (!isNaN(parsed)) {
          return this.compareData(expected, parsed, path, tolerance);
        }
      }
      return [{
        fieldPath: path,
        expectedValue: typeof expected,
        actualValue: typeof actual,
        mismatchType: 'type',
        description: 'Types do not match'
      }];
    }

    // Compare based on type
    if (typeof expected === 'object') {
      if (Array.isArray(expected)) {
        mismatches.push(...this.compareArrays(expected, actual, path, tolerance));
      } else {
        mismatches.push(...this.compareObjects(expected, actual, path, tolerance));
      }
    } else if (typeof expected === 'number') {
      if (Math.abs(expected - actual) > tolerance) {
        mismatches.push({
          fieldPath: path,
          expectedValue: expected,
          actualValue: actual,
          mismatchType: 'value',
          description: `Numeric values differ (tolerance: ${tolerance})`
        });
      }
    } else if (expected !== actual) {
      mismatches.push({
        fieldPath: path,
        expectedValue: expected,
        actualValue: actual,
        mismatchType: 'value',
        description: 'Values are not equal'
      });
    }

    this.mismatches.push(...mismatches);
    return mismatches;
  }

  private compareObjects(
    expected: Record<string, any>,
    actual: Record<string, any>,
    path: string,
    tolerance: number
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    const expectedKeys = new Set(Object.keys(expected));
    const actualKeys = new Set(Object.keys(actual));

    // Check for missing keys
    const missingKeys = Array.from(expectedKeys).filter(key => !actualKeys.has(key));
    for (const key of missingKeys) {
      mismatches.push({
        fieldPath: path ? `${path}.${key}` : key,
        expectedValue: expected[key],
        actualValue: 'missing',
        mismatchType: 'missing',
        description: `Expected key '${key}' is missing`
      });
    }

    // Check for extra keys
    const extraKeys = Array.from(actualKeys).filter(key => !expectedKeys.has(key));
    for (const key of extraKeys) {
      mismatches.push({
        fieldPath: path ? `${path}.${key}` : key,
        expectedValue: 'not expected',
        actualValue: actual[key],
        mismatchType: 'extra',
        description: `Unexpected key '${key}' found`
      });
    }

    // Compare common keys
    for (const key of Array.from(expectedKeys).filter(key => actualKeys.has(key))) {
      const keyPath = path ? `${path}.${key}` : key;
      mismatches.push(...this.compareData(expected[key], actual[key], keyPath, tolerance));
    }

    return mismatches;
  }

  private compareArrays(
    expected: any[],
    actual: any[],
    path: string,
    tolerance: number
  ): DataMismatch[] {
    const mismatches: DataMismatch[] = [];

    if (expected.length !== actual.length) {
      mismatches.push({
        fieldPath: path,
        expectedValue: expected.length,
        actualValue: actual.length,
        mismatchType: 'structure',
        description: 'Array lengths do not match'
      });
      return mismatches;
    }

    for (let i = 0; i < expected.length; i++) {
      const itemPath = `${path}[${i}]`;
      mismatches.push(...this.compareData(expected[i], actual[i], itemPath, tolerance));
    }

    return mismatches;
  }

  /**
   * Log all accumulated mismatches
   */
  logMismatches(correlationId?: string, context?: Record<string, any>): void {
    if (this.mismatches.length === 0) return;

    // Log validation failure as an internal diagnostic event using outbound direction
    logWebSocketMessage(
      'outbound',
      'data_validation',
      {
        mismatchCount: this.mismatches.length,
        mismatches: this.mismatches,
        validationType: 'frontend_data_validation',
        ...context
      },
      {
        correlationId
      }
    );

    if (this.strictMode) {
      throw new Error(`Data validation failed with ${this.mismatches.length} mismatches`);
    }
  }

  /**
   * Clear accumulated mismatches
   */
  clearMismatches(): void {
    this.mismatches = [];
  }

  /**
   * Get all mismatches
   */
  getMismatches(): DataMismatch[] {
    return [...this.mismatches];
  }
}

// Global validator instances
export const defaultValidator = new DataValidator(false);
export const strictValidator = new DataValidator(true);

/**
 * Validate WebSocket message structure
 */
export function validateWebSocketMessage(
  message: any,
  expectedSchema?: any,
  correlationId?: string
): ValidationResult {
  const validator = new DataValidator();

  // Basic WebSocket envelope validation
  const envelopeSchema = {
    type: 'object',
    properties: {
      type: { type: 'string', required: true },
      resource: { type: 'string' },
      data: { type: 'object' },
      timestamp: { type: 'string' },
      correlation_id: { type: 'string' },
      request_id: { type: 'string' }
    }
  };

  let mismatches = validator.validateStructure(message, envelopeSchema);

  // Validate message-specific schema if provided
  if (expectedSchema && message.data) {
    mismatches = mismatches.concat(validator.validateStructure(message.data, expectedSchema, 'data'));
  }

  validator.logMismatches(correlationId, { validationType: 'websocket_message' });

  return {
    isValid: mismatches.length === 0,
    mismatches,
    validatedAt: new Date().toISOString()
  };
}

/**
 * Compare backend data with frontend received data
 */
export function compareBackendFrontendData(
  backendData: any,
  frontendReceivedData: any,
  dataType: string,
  correlationId?: string,
  tolerance: number = 0.001
): ValidationResult {
  const validator = new DataValidator();

  const mismatches = validator.compareData(backendData, frontendReceivedData, '', tolerance);

  if (mismatches.length > 0) {
    validator.logMismatches(correlationId, {
      validationType: 'backend_frontend_comparison',
      dataType,
      backendDataSummary: {
        type: typeof backendData,
        size: JSON.stringify(backendData || '').length
      },
      frontendDataSummary: {
        type: typeof frontendReceivedData,
        size: JSON.stringify(frontendReceivedData || '').length
      }
    });
  }

  return {
    isValid: mismatches.length === 0,
    mismatches,
    validatedAt: new Date().toISOString()
  };
}

/**
 * Validate common message types
 */
export const messageSchemas = {
  signal: {
    type: 'object',
    properties: {
      signal: { type: 'string', enum: ['STRONG_BUY', 'BUY', 'HOLD', 'SELL', 'STRONG_SELL'] },
      confidence: { type: 'number', required: true },
      reasoning_chain: { type: 'object' },
      model_predictions: { type: 'array' },
      model_consensus: { type: 'array' },
      individual_model_reasoning: { type: 'array' },
      timestamp: { type: 'string', required: true }
    }
  },

  portfolio: {
    type: 'object',
    properties: {
      total_value: { type: 'number', required: true },
      available_balance: { type: 'number', required: true },
      positions: { type: 'array' },
      timestamp: { type: 'string' }
    }
  },

  trade: {
    type: 'object',
    properties: {
      order_id: { type: 'string', required: true },
      symbol: { type: 'string', required: true },
      side: { type: 'string', enum: ['buy', 'sell'], required: true },
      quantity: { type: 'number', required: true },
      price: { type: 'number', required: true },
      timestamp: { type: 'string', required: true }
    }
  }
};

File: backend/core/data_validation.py
"""
Data validation utilities for ensuring data consistency between backend and frontend.

Provides utilities to validate data structures, compare expected vs actual data,
and log data mismatches for debugging integration issues.
"""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional, Set, Tuple, Union
from decimal import Decimal
from datetime import datetime, date
import structlog

from backend.core.communication_logger import log_communication

logger = structlog.get_logger()


class DataValidationError(Exception):
    """Raised when data validation fails."""
    pass


class DataMismatch:
    """Represents a data mismatch between expected and actual data."""

    def __init__(
        self,
        field_path: str,
        expected_value: Any,
        actual_value: Any,
        mismatch_type: str,
        description: str = ""
    ):
        self.field_path = field_path
        self.expected_value = expected_value
        self.actual_value = actual_value
        self.mismatch_type = mismatch_type  # "type", "value", "missing", "extra", "structure"
        self.description = description

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for logging."""
        return {
            "field_path": self.field_path,
            "expected_value": self.expected_value,
            "actual_value": self.actual_value,
            "mismatch_type": self.mismatch_type,
            "description": self.description
        }


class DataValidator:
    """Utility for validating data structures and content."""

    def __init__(self, strict_mode: bool = False):
        """Initialize validator.

        Args:
            strict_mode: If True, raise exceptions on validation failures.
                         If False, log warnings instead.
        """
        self.strict_mode = strict_mode
        self.mismatches: List[DataMismatch] = []

    def validate_structure(
        self,
        data: Any,
        expected_schema: Dict[str, Any],
        path: str = ""
    ) -> List[DataMismatch]:
        """Validate data structure against expected schema.

        Args:
            data: The data to validate
            expected_schema: Expected structure/schema
            path: Current path in the data structure (for error reporting)

        Returns:
            List of mismatches found
        """
        mismatches = []

        if expected_schema.get("type") == "object":
            mismatches.extend(self._validate_object_structure(data, expected_schema, path))
        elif expected_schema.get("type") == "array":
            mismatches.extend(self._validate_array_structure(data, expected_schema, path))
        else:
            mismatches.extend(self._validate_primitive(data, expected_schema, path))

        # Check required fields
        if expected_schema.get("required", False) and data is None:
            mismatches.append(DataMismatch(
                path,
                "non-null value",
                None,
                "missing",
                "Required field is null or missing"
            ))

        self.mismatches.extend(mismatches)
        return mismatches

    def _validate_object_structure(
        self,
        data: Any,
        schema: Dict[str, Any],
        path: str
    ) -> List[DataMismatch]:
        """Validate object structure."""
        mismatches = []

        if not isinstance(data, dict):
            return [DataMismatch(
                path,
                "object",
                type(data).__name__,
                "type",
                "Expected object but got different type"
            )]

        expected_properties = schema.get("properties", {})

        # Check for missing required properties
        required_props = schema.get("required", [])
        for prop in required_props:
            if prop not in data:
                mismatches.append(DataMismatch(
                    f"{path}.{prop}" if path else prop,
                    "present",
                    "missing",
                    "missing",
                    f"Required property '{prop}' is missing"
                ))

        # Check existing properties
        for prop, prop_schema in expected_properties.items():
            prop_path = f"{path}.{prop}" if path else prop
            if prop in data:
                mismatches.extend(self.validate_structure(data[prop], prop_schema, prop_path))
            elif prop_schema.get("required", False):
                mismatches.append(DataMismatch(
                    prop_path,
                    "present",
                    "missing",
                    "missing",
                    f"Required property '{prop}' is missing"
                ))

        # Check for extra properties if schema is strict
        if schema.get("additionalProperties", True) is False:
            allowed_props = set(expected_properties.keys())
            actual_props = set(data.keys())
            extra_props = actual_props - allowed_props
            for extra_prop in extra_props:
                prop_path = f"{path}.{extra_prop}" if path else extra_prop
                mismatches.append(DataMismatch(
                    prop_path,
                    "not present",
                    data[extra_prop],
                    "extra",
                    f"Extra property '{extra_prop}' not allowed by schema"
                ))

        return mismatches

    def _validate_array_structure(
        self,
        data: Any,
        schema: Dict[str, Any],
        path: str
    ) -> List[DataMismatch]:
        """Validate array structure."""
        mismatches = []

        if not isinstance(data, list):
            return [DataMismatch(
                path,
                "array",
                type(data).__name__,
                "type",
                "Expected array but got different type"
            )]

        item_schema = schema.get("items", {})

        # Validate array length constraints
        min_items = schema.get("minItems")
        max_items = schema.get("maxItems")

        if min_items is not None and len(data) < min_items:
            mismatches.append(DataMismatch(
                path,
                f"at least {min_items} items",
                len(data),
                "value",
                f"Array has {len(data)} items but minimum is {min_items}"
            ))

        if max_items is not None and len(data) > max_items:
            mismatches.append(DataMismatch(
                path,
                f"at most {max_items} items",
                len(data),
                "value",
                f"Array has {len(data)} items but maximum is {max_items}"
            ))

        # Validate each item
        for i, item in enumerate(data):
            item_path = f"{path}[{i}]"
            mismatches.extend(self.validate_structure(item, item_schema, item_path))

        return mismatches

    def _validate_primitive(
        self,
        data: Any,
        schema: Dict[str, Any],
        path: str
    ) -> List[DataMismatch]:
        """Validate primitive values."""
        mismatches = []

        expected_type = schema.get("type")
        if expected_type:
            actual_type = self._get_json_type(data)

            # Allow numeric types to be flexible (int/float/Decimal)
            if expected_type in ["number", "integer", "float"]:
                if actual_type not in ["number", "integer", "float"]:
                    mismatches.append(DataMismatch(
                        path,
                        expected_type,
                        actual_type,
                        "type",
                        f"Expected {expected_type} but got {actual_type}"
                    ))
            elif expected_type != actual_type:
                mismatches.append(DataMismatch(
                    path,
                    expected_type,
                    actual_type,
                    "type",
                    f"Expected {expected_type} but got {actual_type}"
                ))

        # Validate enum values
        enum_values = schema.get("enum")
        if enum_values and data not in enum_values:
            mismatches.append(DataMismatch(
                path,
                enum_values,
                data,
                "value",
                f"Value {data} not in allowed enum values {enum_values}"
            ))

        return mismatches

    def _get_json_type(self, value: Any) -> str:
        """Get JSON-compatible type name."""
        if value is None:
            return "null"
        elif isinstance(value, bool):
            return "boolean"
        elif isinstance(value, (int, float, Decimal)):
            return "number"
        elif isinstance(value, str):
            return "string"
        elif isinstance(value, list):
            return "array"
        elif isinstance(value, dict):
            return "object"
        else:
            return type(value).__name__

    def compare_data(
        self,
        expected: Any,
        actual: Any,
        path: str = "",
        tolerance: float = 0.0
    ) -> List[DataMismatch]:
        """Compare expected vs actual data for equality.

        Args:
            expected: Expected data
            actual: Actual data
            path: Current path in data structure
            tolerance: Numeric tolerance for floating point comparisons

        Returns:
            List of mismatches found
        """
        mismatches = []

        # Handle None/null values
        if expected is None and actual is None:
            return mismatches
        elif expected is None or actual is None:
            return [DataMismatch(
                path,
                expected,
                actual,
                "value",
                "One value is null, other is not"
            )]

        # Handle different types
        if type(expected) != type(actual):
            # Special case: allow int/float interchange for numeric values
            if isinstance(expected, (int, float, Decimal)) and isinstance(actual, (int, float, Decimal)):
                pass  # Continue with numeric comparison
            else:
                return [DataMismatch(
                    path,
                    type(expected).__name__,
                    type(actual).__name__,
                    "type",
                    "Types do not match"
                )]

        # Compare based on type
        if isinstance(expected, dict):
            mismatches.extend(self._compare_dicts(expected, actual, path, tolerance))
        elif isinstance(expected, list):
            mismatches.extend(self._compare_lists(expected, actual, path, tolerance))
        elif isinstance(expected, (int, float, Decimal)):
            if not self._numeric_equal(expected, actual, tolerance):
                mismatches.append(DataMismatch(
                    path,
                    expected,
                    actual,
                    "value",
                    f"Numeric values differ (tolerance: {tolerance})"
                ))
        elif expected != actual:
            mismatches.append(DataMismatch(
                path,
                expected,
                actual,
                "value",
                "Values are not equal"
            ))

        self.mismatches.extend(mismatches)
        return mismatches

    def _compare_dicts(
        self,
        expected: Dict[str, Any],
        actual: Dict[str, Any],
        path: str,
        tolerance: float
    ) -> List[DataMismatch]:
        """Compare two dictionaries."""
        mismatches = []

        expected_keys = set(expected.keys())
        actual_keys = set(actual.keys())

        # Check for missing keys
        missing_keys = expected_keys - actual_keys
        for key in missing_keys:
            mismatches.append(DataMismatch(
                f"{path}.{key}" if path else key,
                expected[key],
                "missing",
                "missing",
                f"Expected key '{key}' is missing"
            ))

        # Check for extra keys
        extra_keys = actual_keys - expected_keys
        for key in extra_keys:
            mismatches.append(DataMismatch(
                f"{path}.{key}" if path else key,
                "not expected",
                actual[key],
                "extra",
                f"Unexpected key '{key}' found"
            ))

        # Compare common keys
        for key in expected_keys & actual_keys:
            key_path = f"{path}.{key}" if path else key
            mismatches.extend(self.compare_data(expected[key], actual[key], key_path, tolerance))

        return mismatches

    def _compare_lists(
        self,
        expected: List[Any],
        actual: List[Any],
        path: str,
        tolerance: float
    ) -> List[DataMismatch]:
        """Compare two lists."""
        mismatches = []

        if len(expected) != len(actual):
            mismatches.append(DataMismatch(
                path,
                len(expected),
                len(actual),
                "structure",
                "List lengths do not match"
            ))
            return mismatches

        for i, (exp_item, act_item) in enumerate(zip(expected, actual)):
            item_path = f"{path}[{i}]"
            mismatches.extend(self.compare_data(exp_item, act_item, item_path, tolerance))

        return mismatches

    def _numeric_equal(self, a: Union[int, float, Decimal], b: Union[int, float, Decimal], tolerance: float) -> bool:
        """Check if two numeric values are equal within tolerance."""
        return abs(float(a) - float(b)) <= tolerance

    def log_mismatches(
        self,
        correlation_id: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> None:
        """Log all accumulated mismatches."""
        if not self.mismatches:
            return

        # Log summary
        log_communication(
            direction="internal",
            protocol="validation",
            message_type="data_validation",
            correlation_id=correlation_id,
            payload={
                "mismatch_count": len(self.mismatches),
                "mismatches": [m.to_dict() for m in self.mismatches]
            },
            target="system",
            error="Data validation failed" if self.strict_mode else None,
            **(context or {})
        )

        # Raise exception in strict mode
        if self.strict_mode:
            raise DataValidationError(f"Data validation failed with {len(self.mismatches)} mismatches")

    def clear_mismatches(self) -> None:
        """Clear accumulated mismatches."""
        self.mismatches.clear()


# Global validator instances
default_validator = DataValidator(strict_mode=False)
strict_validator = DataValidator(strict_mode=True)


def validate_websocket_message(
    message: Dict[str, Any],
    expected_schema: Optional[Dict[str, Any]] = None,
    correlation_id: Optional[str] = None
) -> List[DataMismatch]:
    """Validate a WebSocket message structure.

    Args:
        message: WebSocket message to validate
        expected_schema: Expected schema for the message
        correlation_id: Correlation ID for logging

    Returns:
        List of validation mismatches
    """
    validator = DataValidator()

    # Basic WebSocket envelope validation
    envelope_schema = {
        "type": "object",
        "properties": {
            "type": {"type": "string", "required": True},
            "resource": {"type": "string"},
            "data": {"type": "object"},
            "timestamp": {"type": "string"},
            "correlation_id": {"type": "string"},
            "request_id": {"type": "string"}
        }
    }

    mismatches = validator.validate_structure(message, envelope_schema)

    # Validate message-specific schema if provided
    if expected_schema and "data" in message:
        data_mismatches = validator.validate_structure(message["data"], expected_schema, "data")
        mismatches.extend(data_mismatches)

    validator.log_mismatches(correlation_id, {"validation_type": "websocket_message"})
    return validator.mismatches


def compare_backend_frontend_data(
    backend_data: Any,
    frontend_received_data: Any,
    data_type: str,
    correlation_id: Optional[str] = None,
    tolerance: float = 0.001
) -> List[DataMismatch]:
    """Compare data sent by backend with what frontend received.

    Args:
        backend_data: Data sent by backend
        frontend_received_data: Data received by frontend
        data_type: Type of data being compared (e.g., "portfolio", "signal")
        correlation_id: Correlation ID for tracking
        tolerance: Numeric tolerance for comparisons

    Returns:
        List of data mismatches
    """
    validator = DataValidator()

    mismatches = validator.compare_data(
        backend_data,
        frontend_received_data,
        tolerance=tolerance
    )

    if mismatches:
        validator.log_mismatches(correlation_id, {
            "validation_type": "backend_frontend_comparison",
            "data_type": data_type,
            "backend_data_summary": {
                "type": type(backend_data).__name__,
                "size": len(json.dumps(backend_data, default=str)) if backend_data else 0
            },
            "frontend_data_summary": {
                "type": type(frontend_received_data).__name__,
                "size": len(json.dumps(frontend_received_data, default=str)) if frontend_received_data else 0
            }
        })

    return mismatches